src/rust/engine/nailgun/src/server.rs

```diff
@@ -233,24 +234,37 @@ impl Nail for RawFdNail {
   ) -> Result<nails::server::Child, io::Error> {
     let env = cmd.env.iter().cloned().collect::<HashMap<_, _>>();
 
-    // Handle stdin.
+    // Handle stdin. If the input stream closes, the run is cancelled.
+    let cancelled = AsyncLatch::new();
     let (stdin_handle, stdin_sink) = Self::input(Self::ttypath_from_env(&env, 0))?;
-    let accepts_stdin = if let Some(mut stdin_sink) = stdin_sink {
-      let mut input_stream = input_stream.filter_map(|child_input| {
-        Box::pin(async move {
-          match child_input {
-            ChildInput::Stdin(bytes) => Some(Ok(bytes)),
-            ChildInput::StdinEOF => None,
+    let accepts_stdin = {
+      let (accepts_stdin, input_stream_fut) = if let Some(mut stdin_sink) = stdin_sink {
```

Comment:
I've never seen an `if let` with two equal signs like this. How is this evaluated?

Reply:
If you ignore the `let` detail, this is just:
```rust
let x = (if ...) // x is assigned the result of the if expression as a whole
```

The let detail is scoped inside the `if` expression and is being used like the walrus operator in python.